ovs: Fix `is_ovs_running()` in container environment.

cathay4t opened the pull request:
In k8s container environment, the OVS database socket
/var/run/openvswitch/db.sock is mounted from host, so NM can managed it
without the ovs daemon running in container.

To support that, this patch removed the top level checking on
`is_ovs_running()` and trust plugin raise the proper error on failure.

Patched the NM plugin to check the error
`NM.DeviceStateReason.OVSDB_FAILED` on activation failure, raise
`NmstateDependencyError` if OVS DB failed to connected.

NM will not raise any error when creating OVS internal interface with
OVSDB mounted to /dev/null, NM will keep showing the OVS interface as
ACTIVATING, changed the fallback checker to give only 30 seconds for OVS
interface to exit `NM.DeviceState.PREPARE`, if not treat it as OVS
daemon malfunctioning.

Updated integration test case to mask(mount /dev/null) the OVS DB socket
file for simulating the stopped OVS daemon.

qinqon commented on libnmstate/ifaces/ovs.py:
```diff
@@ -255,7 +255,7 @@ def _validate_ovs_mtu_mac_confliction(self):
     try:
         subprocess.run(
-            ("systemctl", "status", "openvswitch"),
+            ("ovsdb-client", "list-dbs"),
```
This will add a dependency on containers, can we use the python OVS client to do same operation ? similar tu the stuff under plugin/ovs...

cathay4t replied:
New approach uploaded.

Instead of doing pre-check, I removed the `is_ovs_running()` and trust nmstate-NM plugin raise the proper exception when OVSDB malfunctioned.

Introducing new dependency to nmstate or nmstate-NM plugin is not ideal in this case, the `ovsdb` plugin is not install by default hence it can hold extra dependency.

qinqon commented on tests/integration/ovs_test.py:
```diff
@@ -214,7 +214,7 @@ def test_nm_ovs_plugin_missing():
 
 
 def test_ovs_service_missing():
```
Let's xfail for k8s this test and fix it later on at different PR stopping ovs servie at nodes and starting it up at teardown.

cathay4t replied:
No need. New patch is using `mount -o bind /dev/null /var/run/openvswitch/db.sock` to simulate the OVSDB daemon stopped.

qinqon commented:
This approach looks better for k8s env but there is still some k8s test failing (just one)

tests.integration.examples_test: test_add_remove_ovs_bridge https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/nmstate_nmstate/1583/pull-nmstate-integ_tier1-k8s/1391604886155563008

It's failing at the new error related to NM not able to mount ovs db socket.

```
raise tmp_error
E           libnmstate.error.NmstateDependencyError: Activate profile uuid:3c50eae4-8361-4753-a0e8-8ceb65b2b25e iface:ovs0 type: ovs-interface failed: timeout on creating OVS interface, please check Open vSwitch daemon

```

cathay4t commented:
@qinqon I failed to reproduce this problem in my VM. Any guide on running k8s test locally?

ffmancera commented:
@cathay4t I am investigating this specific bug on Nmstate. If it takes too long I will mark it as xfail for k8s and put the effort later. Thank you!

qinqon commented:
> @qinqon I failed to reproduce this problem in my VM. Any guide on running k8s test locally?

This is how is done on the prow job https://github.com/kubevirt/project-infra/blob/master/github/ci/prow-deploy/files/jobs/nmstate/nmstate/nmstate-presubmits.yaml#L32

```
automation/run-tests.sh --test-type integ_tier1 --centos-stream --k8s
```

You can use `--debug-shell` too

qinqon commented:
> @cathay4t I am investigating this specific bug on Nmstate. If it takes too long I will mark it as xfail for k8s and put the effort later. Thank you!

Let's see if @cathay4t can reproduce it and if we see it's going to take long time to fix we mark it as xfail.

cathay4t committed:
ovs: Fix `is_ovs_running()` in container environment.

In k8s container environment, the OVS database socket
/var/run/openvswitch/db.sock is mounted from host, so NM can managed it
without the ovs daemon running in container.

To support that, this patch removed the top level checking on
`is_ovs_running()` and trust plugin raise the proper error on failure.

Patched the NM plugin to check the error
`NM.DeviceStateReason.OVSDB_FAILED` on activation failure, raise
`NmstateDependencyError` if OVS DB failed to connected.

NM will not raise any error when creating OVS internal interface with
OVSDB mounted to /dev/null, NM will keep showing the OVS interface as
ACTIVATING, changed the fallback checker to give only 30 seconds for OVS
interface to exit `NM.DeviceState.PREPARE`, if not treat it as OVS
daemon malfunctioning.

Updated integration test case to mask(mount /dev/null) the OVS DB socket
file for simulating the stopped OVS daemon.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t commented:
I failed to debug out the problem. Since the test case has been marked as xfail in k8s, please proceed on reviewing this PR.

I may continue the investigation later.

qinqon commented:
> I failed to debug out the problem. Since the test case has been marked as xfail in k8s, please proceed on reviewing this PR.
> 
> I may continue the investigation later.

There is a new tests failing becaus the xfail is strict and the container NetworkManager version is older than the one running at node, don't we have to bump NetworkManager version a nmstate-dev container ? 

The centos-stream-nmstate-dev image has a quite old NetworkManager version 

```
[ellorent@localhost nmstate]$ docker run docker.io/nmstate/centos-stream-nmstate-dev rpm -qa |grep NetworkManager
NetworkManager-libnm-1.26.0-12.el8_3.x86_64
NetworkManager-ovs-1.26.0-12.el8_3.x86_64
NetworkManager-1.26.0-12.el8_3.x86_64
NetworkManager-team-1.26.0-12.el8_3.x86_64
NetworkManager-config-server-1.26.0-12.el8_3.noarch
```

UPDATE: After talking @ffmancera looks like the centos-stream-nmstate-dev container need to be updated so we use the latest NetworkManager from centos-stream wich is 1.32.

cathay4t commented:
@qinqon Centos stream container updated and your CI passed.

qinqon commented:
@cathay4t all good you can merge this.

ffmancera reviewed (approved):
LGTM, thanks!
